### ISSUE TITLE
Proposal: addition of new test `has_object()`

### ISSUE THREAD

**Issue by aroberge** [aroberge/reeborg#106]
Tokens play a special role in Reeborg's World; they are similar to beepers in the original Karel the robot world which were the only objects.   There are two commands associated with tokens and two tests:

    take()
    put()
    token_here()
    has_token()

Reeborg's world is richer than Karel in that other type of objects may be possible (and, indeed, we are considering adding even more objects).  To the above four functions, we have the following:

    take('object_name')
    put('object_name')
    object_here()   # returns False or the type of object here - there can only be one type at a given location

So, there is no corresponding test to `has_token()`.   I propose to add `has_object()` which will return True if Reeborg carries anything, and False otherwise.   To be more specific, Reeborg can use `has_object("object_name")` which will return True if Reeborg carries at least one object of the specified type, and False otherwise.

**Comment by Goldsong** [comment]
I have not had the need to query the robot about what it the type of object that it carries.  I sometimes use the star as a goal, in which case the user has to explicitly request to take the star with take("star"). Although this has not been a problem so far, take() is an issue because is the only command in Reeborg that can be called with and without arguments:

take()  # takes a token
take("star")  # takes a star

Nobody has pointed this yet, which is good because a beginner's class is not the place to discuss either variable-length arguments or default arguments. It would be better if take() would never take arguments and instead, simply picks up whatever object happens to be on the cell, be it a token or other, with the same take() command.

W.r.t. the query functions, object_here() serves the same purpose as token_here() when there are tokens present but also works if there are other objects, like stars or squares. Likewise, has_tokens() could be replaced with the neutral has_object().

With these functionality, there is really no difference between a token and a star. however, at this moment Reeborg treats tokens different from stars, squares and triangles, because there may be stacks of tokens but there can only be a single star, square or rectangle per cell. How about removing this restriction, i.e. allow stacks of 'objects'. Whether the visual representation of the 'object' is a token, or star, or triangle, or strawberry, is secondary and could be selected from a menu or grid, similar to the way that now we ware selecting the objects.

**Comment by aroberge** [comment]
I agree with the idea of having more than one object of a given type at a given location, like we do for token.

I require the use of the return value for `object_here()` for some missions (see http://reeborg.ca/docs/begin_py_en/harvest3.html) These would be more interesting if we had images of fruits instead of just geometrical shapes...      I find these a good example to demonstrate the effect of the `return` keyword.

With the current restriction of having only one object of a given type at a given location (like we currently have), I see the point in not having any argument for `take()` i.e. if only one object is present, `take()` will automatically refer to that object.   With 1) this modification, and 2) with the possibility of having more than one object of a given type at a certain grid location, and 3) having many different types of objects with nice images, we no longer would have to use tokens as the default object to simulate whatever Reeborg needs to do.    Thus, for beginners, if we want to have Reeborg harvest strawberries, (http://codeperspectives.com/intro/learn-program-libraries/) we could use images of strawberries and still have a program that use functions without arguments.    I do like the idea of this modification.

However, while this change would be helpful to design more interesting worlds for beginners, I am still interested in having Reeborg's World be used to demonstrate more advanced concepts, like the use of default arguments.    I can think of maintaining the restriction of not allowing Reeborg to `put()` one object at a location if an object of a different type is present.  However, I could see having worlds where more than one object type (e.g. weed and strawberry) are potentially present at a given location, and having Reeborg take only one of them. [this is not currently possible, but should be - the challenge will be to have the right images....]

To summarize my thinking: **yes** to make it simpler for beginners to deal with interesting worlds (e.g. `take()` without argument works with whatever object is there), but keep the flexibility of using more complex Python constructs (e.g. `take(argument)`) for more advanced students.

**Comment by aroberge** [comment]
Both `object_here()` and `has_object()` have been implemented using the same semantics.  **Before I describe it, I'm wondering if `has_object()` should not be named** `carries()`.

For `object_here()`: 

1. if there are no objects in the world, at the location where the robot is, the value returned is `False`.
2. if  there are objects at that location, their names is returned as a list.  For example, suppose that there are three stars and two tokens at that location, `object_here()` would return `["star", "token"]`  (not necessarily in that order.   Note that such a list is treated as a **true** value in either Python or Javascript, so that `if object_here():` should behave as expected.

For `object_here(name)`:

1. if there are no such object at that location, the value returned is `False`
2. if there is at least one such object at that location, the value returned is `True`.

Somewhat related ....

For `take()`:

1. if there is only one type of object at that location, one object of that type will be picked up -- this is essentially how it worked for tokens, but now it will work that way for all types of objects.
2. if there are two different types of objects at that location, an error will be raised as Reeborg cannot decide which object to take.  In this case we need to specify `take(name)`.

`has_object()` follows the same logic as `object_here()` except that it is based on the objects carried by the robot.

`put()` follows the same logic as `take()` except that it is based on the objects carried by the robot.

Thus, for beginners, as long as there is only one type of objects in the world (and, if they are not tokens, ensure that the world is created so that the robot does not carry any tokens), `take()` and `put()` can be used without worrying about specifying an argument type.

**Comment by Goldsong** [comment]
carries_object() is very easy to remember; it would work very well. All the new commands are backwards compatible.

**Comment by aroberge** [comment]
On Tue, May 26, 2015 at 11:14 AM, Andres Castano <notifications@github.com>
wrote:

> carries_object() is very easy to remember; it would work very well. All
> the new commands are backwards compatible.
>
>

I will change it right away.
